src/v/net/transport.cc

```diff
@@ -123,4 +123,10 @@ void base_transport::shutdown() noexcept {
     }
 }
 
+ss::future<> base_transport::wait_input_shutdown() {
+    if (_fd && _shutdown) {
```

Comment:
I don't quite understand why do we need to introduce `_shutdown` variable.
The `wait_for_input_shutdown` is only called from `stop` method. So it will be always set. Also, the var is never reset back to `false`.

Reply:
This code is pretty convoluted. I got lost while looking at the diff and I wrote it.

> I don't quite understand why do we need to introduce _shutdown variable.

Previously, `_fd` would be reset every time `base_transport::shutdown()` was called. We want to call,
`wait_input_shutdown` after that point, so we can't reset the pointer there. `_fd` being null also
implied that the connection was not valid anymore (see `base_transport::is_valid`). Putting these together, if we didn't have `_shutdown` we would always think the connection is valid. In short, `_shutdown` is there to keep `_fd` alive for longer while preserving the same semantics for the callers.

> The wait_for_input_shutdown is only called from stop method. So it will be always set.

There's no guarantee this will always be the case though.

> Also, the var is never reset back to false.

We reset it back to `false` when a new connection is created. See the call to `base_transport::reset_state` in `base_transport::do_connect`.

Reply:
OK, got it. I missed the reset_state bit.